feat: sign linux system packages with GPG identities (#2973)
beeware/briefcase: docs/en/how-to/code-signing/linux.md

# Linux { #code-signing-linux }

## Overview

> Linux system packages (`.deb`, `.rpm` and `.pkg.tar.zst`) can be signed using a [GNU Privacy Guard](https://gnupg.org/) (GPG) key. Unlike macOS and Windows, there is no central authority that issues signing certificates for Linux. A GPG signature does not, by itself, make a package trusted; the person installing your package must have imported your public key into their GPG keyring, and must choose to trust it. However, a signature does allow a user to verify both the *integrity* and the *provenance* of the package, and it is a requirement for publishing packages to some official repositories.

Reviewer: No need for the emphasis here.
```suggestion
Linux system packages (`.deb`, `.rpm` and `.pkg.tar.zst`) can be signed using a [GNU Privacy Guard](https://gnupg.org/) (GPG) key. Unlike macOS and Windows, there is no central authority that issues signing certificates for Linux. A GPG signature does not, by itself, make a package trusted; the person installing your package must have imported your public key into their GPG keyring, and must choose to trust it. However, a signature does allow a user to verify both the integrity and the provenance of the package, and it is a requirement for publishing packages to some official repositories.
```

Author: thanks, agreed. i'll apply your suggested wording.

## Create a GPG key

If you don't already have a GPG key, you can create one with:

```console
$ gpg --full-generate-key
```

> You will be prompted to select a key type, key size and expiry date, and to provide a name and email address that will identify the key. The email address should be an address you control, as users will use it (along with your public key) to identify that the package really came from you.

Reviewer: When I ran this, I was presented with a range of options for key type - can you provide any meaningful guidance about what type of key should be generated? On Ubuntu and Fedora, it suggests "ECC (sign and encrypt)" as a default - is that reasonable? What about the other options?

Author: Yes, "ECC (sign and encrypt)" is the right default. Since GnuPG 2.4, ECC (Ed25519 for signing + Curve 25519 for encryption) is GnuPG's own recommended default,  it's fast, produces small signatures, and is what current GnuPG versions on Ubuntu/Fedora will suggest. I'll add a recommendation for that to the docs.

Of the other options: RSA (4096 bit) is the compatibility fallback — it works with much older GnuPG versions and older verification tooling (including `rpmsign`/`debsigs` on older distributions), at the cost of larger/slower keys. The other ECC curves (NIST P-384, Brainpool) offer no practical advantage for package signing, and DSA is long since deprecated, so I won't recommend those.

```console
$ gpg --list-secret-keys
```

Each key is listed with a `sec` entry, followed by the key's 40-character hexadecimal fingerprint. This fingerprint is the value you will provide as the `identity` for the Briefcase package command:

```console
$ briefcase package linux system --identity <fingerprint>
```

Alternatively, you can specify any portion of the fingerprint, the short key ID, or the name or email address associated with the key. If you have more than one secret key on your system, and you don't specify an identity, Briefcase will prompt you to select the identity to use.

## Distributing the signature

When Briefcase signs a package, the signature is stored in a file alongside the package:

- A `.deb` package is signed in place, and the signature is embedded in the package's `origin` signature field.
- An `.rpm` package is signed in place; the signature is embedded in the package header.
- An `.pkg.tar.zst` package is signed with a detached signature, producing a separate `<package>.sig` file that must be distributed alongside the package.

> Before distributing your packages, you should export your public key and make it available so that users can verify your packages:

Reviewer: What does "make it available" mean? Where/how should that be done?

Author: fair,  that was vague. I'll clarify that you should either publish the key to a keyserver (e.g.,` gpg --send-keys --keyserver keys.openpgp.org <fingerprint>`) or host the exported `.asc` file somewhere users can fetch it, and that users then import it into their keyring. I'll spell both out.

```console
$ gpg --export --armor <fingerprint>
```

> This will output your public key in ASCII armor format, which can be added to a user's keyring with `gpg --import`. A user can then verify the integrity and provenance of your package with `gpg --verify <signature file> <package file>` (for `.pkg.tar.zst` packages), or with the package manager's own verification tooling (for `.deb` and `.rpm` packages).

Reviewer: Does "with the package managers own verification tooling" mean that the signature is automatically checked when the package is installed? Or that there's an option to `apt`/`dnf` that can verify a package?

Author: The signature is not automatically verified when the package is installed — I'll reword this to be explicit. The embedded signature on .deb/.rpm packages is checked with the dedicated tooling (`debsig-verify` for `.deb`, `rpmkeys --checksig/rpm --checksig` for` .rpm`), while `.pkg.tar.zst` signatures are verified with `gpg --verify`. I'll correct the doc so it doesn't imply automatic verification.

# Linux { #code-signing-linux }

## Overview

Linux system packages (`.deb`, `.rpm` and `.pkg.tar.zst`) can be signed using a [GNU Privacy Guard](https://gnupg.org/) (GPG) key. Unlike macOS and Windows, Linux has no central authority that issues signing certificates. A GPG signature lets users verify the integrity and provenance of a package, but it does not by itself make a package trusted; the person installing your package must import your public key into their GPG keyring and choose to trust it. A signature is required for publishing packages to some official repositories.

## Create a GPG key

If you don't already have a GPG key, you can create one with:

```console
$ gpg --full-generate-key
```

> You will be prompted to select a key type, key size and expiry date, and to provide a name and email address that will identify the key. If possible, use an ECC key based on Curve 25519 (an `ed25519` signing key), which is the default in recent GnuPG versions and produces smaller, faster signatures; if you need to support older tools that don't understand ECC keys, generate an RSA key of at least 4096 bits instead. The email address should be an address you control, as users will use it (along with your public key) to identify that the package really came from you.

Reviewer: ```suggestion
You will be prompted to select a key type, key size and expiry date, and to provide a name and email address that will identify the key. If possible, use an ECC key based on Curve 25519 (an `ed25519` signing key), which is the default in recent GnuPG versions and produces smaller, faster signatures. If you need to support older tools that don't understand ECC keys, generate an RSA key of at least 4096 bits instead. The email address should be an address you control, as users will use it (along with your public key) to identify that the package really came from you.
```

```console
$ briefcase package linux system --identity <fingerprint>
```

Alternatively, you can specify any portion of the fingerprint, the short key ID, or the name or email address associated with the key. If you don't specify an identity, Briefcase will prompt you to select the identity to use from the secret keys available on your system, or to opt out of signing. If you have only one secret key, it is offered as the default selection.

## Distributing the signature

When Briefcase signs a package, the signature is stored in a file alongside the package:

- A `.deb` package is signed in place, and the signature is embedded in the package's `origin` signature field.
- An `.rpm` package is signed in place; the signature is embedded in the package header.
- An `.pkg.tar.zst` package is signed with a detached signature, producing a separate `<package>.sig` file that must be distributed alongside the package.

Before distributing your packages, you should export your public key and make it available so that users can verify your packages:

```console
$ gpg --export --armor <fingerprint>
```

> This will output your public key in ASCII armor format. You can publish it on a public keyserver, for example with `gpg --send-keys --keyserver keys.openpgp.org <fingerprint>`, or you can host the exported `.asc` file on your own website. Users can then add your key to their keyring with `gpg --import <file>` (or `gpg --recv-keys <fingerprint>`).

Reviewer: ```suggestion
This will output your public key in ASCII armor format. You can publish it on a public keyserver (e.g., with `gpg --send-keys --keyserver keys.openpgp.org <fingerprint>`), or you can host the exported `.asc` file on your own website. Users can then add your key to their keyring with `gpg --import <file>` (or `gpg --recv-keys <fingerprint>`).
```

> A user can then verify the integrity and provenance of your package: with `gpg --verify <signature file> <package file>` for `.pkg.tar.zst` packages; with `debsig-verify` for `.deb` packages (using a policy that declares your key as trusted); or with `rpmkeys --checksig` (or `rpm --checksig`) for `.rpm` packages, once RPM is configured to trust your key.

Reviewer: ```suggestion
A user can then verify the integrity and provenance of your package: 

- `debsig-verify` for `.deb` packages (using a policy that declares your key as trusted).
- `rpmkeys --checksig` (or `rpm --checksig`) for `.rpm` packages, once RPM is configured to trust your key.
- `gpg --verify <signature file> <package file>` for `.pkg.tar.zst` packages.
```
